flink-ml-lib/src/main/java/org/apache/flink/ml/recommendation/Swing.java

```diff
@@ -0,0 +1,407 @@
+			super.snapshotState(context);
+			userItemsMapState.update(Collections.singletonList(userItemsMap));
+		}
+	}
+
+	/**
+	 * Calculate top N similar items of each item.
+	 */
+	private static class CalculateSimilarity
+		extends AbstractStreamOperator <Row>
+		implements OneInputStreamOperator <
+		Tuple3 <String, String, List <String>>,
+		Row>,
+		BoundedOneInput {
+
+		private Map <String, HashSet <String>> userItemsMap = new HashMap <>();
+		private Map <String, HashSet <String>> itemUsersMap = new HashMap <>();
+		private ListState <Map <String, List <String>>> userItemsMapState;
+		private ListState <Map <String, List <String>>> itemUsersMapState;
+
+		final private double alpha = 1.0;
+		final private double userAlpha = 5.0;
+		final private double userBeta = -0.35;
+		final private int topN;
+
+		private CalculateSimilarity(int topN) {this.topN = topN;}
+
+		@Override
+		public void endInput() throws Exception {
+
+			Map <String, Float> userWeights = new HashMap <>(userItemsMap.size());
+			userItemsMap.forEach((k, v) -> {
+				int count = v.size();
+				userWeights.put(k, calculateWeight(count));
+			});
+
+			for (String mainItem : itemUsersMap.keySet()) {
+				List <String> userList = new ArrayList(itemUsersMap.get(mainItem));
+				HashMap <String, Float> id2swing = new HashMap <>();
+
+				for (int i = 0; i < userList.size(); i++) {
+					String u = userList.get(i);
+					for (int j = i + 1; j < userList.size(); j++) {
+						String v = userList.get(j);
+						HashSet <String> interaction = (HashSet <String>) userItemsMap.get(u).clone();
+						interaction.retainAll(userItemsMap.get(v));
+						if (interaction.size() == 0) {
+							continue;
+						}
+						float similarity =
+							(float) (userWeights.get(u) * userWeights.get(v) / (alpha + interaction.size()));
+						for (String simItem : interaction) {
+							if (simItem.equals(mainItem)) {
+								continue;
+							}
+							float itemSimilarity = id2swing.getOrDefault(simItem, (float) 0) + similarity;
+							id2swing.put(simItem, itemSimilarity);
+						}
+					}
+				}
+
+				ArrayList <Tuple2 <String, Float>> itemAndScore = new ArrayList <>();
+				id2swing.forEach(
+					(key, value) -> itemAndScore.add(Tuple2.of(key, value))
+				);
+
+				itemAndScore.sort(new Comparator <Tuple2 <String, Float>>() {
+					@Override
+					public int compare(Tuple2 <String, Float> o1, Tuple2 <String, Float> o2) {
+						return 0 - Float.compare(o1.f1, o2.f1);
+					}
+				});
+
+				if (itemAndScore.size() == 0) {
+					continue;
+				}
+
+				int itemNums = topN > itemAndScore.size() ? itemAndScore.size() : topN;
+				String[] itemIds = new String[itemNums];
+				Float[] itemScores = new Float[itemNums];
+				for (int i = 0; i < itemNums; i++) {
+					itemIds[i] = itemAndScore.get(i).f0;
+					itemScores[i] = itemAndScore.get(i).f1;
+				}
+
+				output.collect(
+					new StreamRecord <>(
+						Row.of(
+							mainItem, new ArrayList <String>(Arrays.asList(itemIds)),
```

Comment:
Thanks for reminding me. I met a `ClassCastException` when use `Arrays.asList(itemIds)` so used an inappropriate class transformation.
> Caused by: java.lang.ClassCastException: java.util.Arrays$ArrayList cannot be cast to java.util.ArrayList
	at org.apache.flink.ml.recommendation.swing.Swing$1.map(Swing.java:130)
	
I have modified type of `itemIds` and `itemScores` to ArrayList, you can check it.

Reply:
I just pulled the code, and tried to modify and run it. 

Actually, you met a `ClassCastException` here because `items` and `scores` in `SwingModelData` are defined as `ArrayList`. After defining them as `ArrayList`, an additional cast is used in `ModelDataDecoder#read`.

At least, the cast in `ModelDataDecoder#read` is inappropriate, as usually you cannot assume more information than the API provides.
For other places, using interfaces usually gives more flexibility than using concrete classes.